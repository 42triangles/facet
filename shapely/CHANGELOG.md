# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.1.0](https://github.com/bearcove/shapely/compare/shapely-v3.0.0...shapely-v3.1.0) - 2025-03-31

### Other

- Fill in missing docs
- Document more stuff
- Improve naming
- More notes
- Fix all markdown links
- More docs
- unfuck docs
- Mhh it derives _something_
- Tuple layouts
- more tuple support
- pre-commit hook with cargo-fmt
- More complex derives work
- Real-life derive, which fails
- Support more attribute
- more derive
- shill for namespace, closes #36
- just pre-commit, just fmt
- set up cargo-husky to run cargo fmt
- specific toolchains, reformat code

## [3.0.0](https://github.com/bearcove/shapely/compare/shapely-v2.0.1...shapely-v3.0.0) - 2025-03-11

### Other

- Fix doc tests in README.md
- Document how to write your deserializer a little better

## [2.0.0](https://github.com/bearcove/shapely/compare/shapely-v1.0.0...shapely-v2.0.0) - 2025-03-11

### Other

- Change Shape.name from static str to NameFn
- Looking good!
- clean up drop impls
- Stability notes
- link to 'free of syn' campaign
- Make derive unconditional, closes #8
- Get rid of insta, closes #10
- Ensure no syn dependency (and badge about it), closes #9
- Introduce core crate
- Get rid of debug/display, closes #4
- Start implementing transparent
