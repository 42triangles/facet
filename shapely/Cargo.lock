# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 4

[[package]]
name = "shapely-core"
version = "0.1.0"
