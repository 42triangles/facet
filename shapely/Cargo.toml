[package]
name = "shapely"
version.workspace = true
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
license.workspace = true
repository.workspace = true
homepage.workspace = true
documentation.workspace = true
description.workspace = true
keywords.workspace = true
categories.workspace = true

[dependencies]
shapely-trait.workspace = true
shapely-derive.workspace = true

[features]
# Does nothing, only used for tests
testfeat = []

[dev-dependencies]
shapely-spez = { path = "../shapely-spez" }
