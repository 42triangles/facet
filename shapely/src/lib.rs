//! Provides the core traits for shapely

#[cfg(feature = "derive")]
pub use shapely_derive::*;

pub use nonmax::NonMaxU32;

use std::alloc;

mod hashmap_impl;
mod scalar_impls;

mod shape;
pub use shape::*;

mod slot;
pub use slot::Slot;

mod uninit;
pub use uninit::*;

#[cfg(all(test, feature = "derive"))]
mod derive_tests;

/// Provides reflection so you can shapely about your types.
pub trait Shapely {
    /// Returns the shape of this type
    fn shape() -> Shape;

    fn uninit() -> ShapeUninit {
        let shape = Self::shape();
        let layout = alloc::Layout::from_size_align(shape.size, shape.align).unwrap();
        let addr = unsafe { alloc::alloc(layout) };
        if addr.is_null() {
            alloc::handle_alloc_error(layout);
        }
        ShapeUninit {
            addr,
            init_fields: InitFields64::new(),
            shape,
        }
    }
}

#[cfg(test)]
mod tests {
    use nonmax::NonMaxU32;

    use crate::{Innards, Shape, Shapely, StructManipulator};

    #[derive(Debug, PartialEq, Eq)]
    struct FooBar {
        foo: u64,
        bar: String,
    }

    impl Shapely for FooBar {
        fn shape() -> crate::Shape {
            use crate::{Innards, MapField, MapInnards};

            fn shape_of<TStruct, TField: Shapely>(_f: impl Fn(TStruct) -> TField) -> Shape {
                TField::shape()
            }
            macro_rules! map_field {
                ($struct:ty, $field:ident) => {
                    MapField {
                        name: stringify!($field),
                        schema: || shape_of(|s: $struct| s.$field),
                        offset: Some(
                            NonMaxU32::new(std::mem::offset_of!($struct, $field) as u32).unwrap(),
                        ),
                    }
                };
            }
            static FOO_FIELD: MapField = map_field!(FooBar, foo);
            static BAR_FIELD: MapField = map_field!(FooBar, bar);

            static SCHEMA: Shape = Shape {
                name: "FooBar",
                size: std::mem::size_of::<FooBar>(),
                align: std::mem::align_of::<FooBar>(),
                innards: Innards::Map(
                    MapInnards::builder()
                        .field(FOO_FIELD)
                        .field(BAR_FIELD)
                        .open_ended(false)
                        .mk_slots(|shape| &StructManipulator { shape })
                        .build(),
                ),
                display: None,
                debug: None,
                set_to_default: None,
            };
            SCHEMA
        }
    }

    #[test]
    fn build_foobar_through_reflection() {
        let schema = FooBar::shape();

        let layout = std::alloc::Layout::from_size_align(schema.size, schema.align).unwrap();
        let ptr = unsafe { std::alloc::alloc(layout) };
        if ptr.is_null() {
            std::alloc::handle_alloc_error(layout);
        }

        if let Innards::Map(sh) = &schema.innards {
            let foo_bar = unsafe { &mut *(ptr as *mut FooBar) };
            for field in sh.fields {
                unsafe {
                    match field.name {
                        "foo" => {
                            if let Some(slot) = sh.slots.slot(foo_bar, *field) {
                                slot.fill(42u64);
                            }
                        }
                        "bar" => {
                            if let Some(slot) = sh.slots.slot(foo_bar, *field) {
                                slot.fill(String::from("Hello, World!"));
                            }
                        }
                        _ => panic!("Unknown field: {}", field.name),
                    }
                }
            }
        }

        // Verify the fields were set correctly
        let foo_bar = unsafe { &*(ptr as *const FooBar) };
        assert_eq!(foo_bar.foo, 42);
        assert_eq!(foo_bar.bar, "Hello, World!");

        assert_eq!(
            &FooBar {
                foo: 42,
                bar: "Hello, World!".to_string()
            },
            foo_bar
        )
    }
}
