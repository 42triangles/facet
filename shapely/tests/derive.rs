use insta::assert_debug_snapshot;
use shapely::Shapely;

#[test]
fn simple_struct() {
    #[derive(Debug, Shapely)]
    struct Blah {
        foo: u32,
        bar: String,
    }

    assert_debug_snapshot!(Blah::shape());
}
