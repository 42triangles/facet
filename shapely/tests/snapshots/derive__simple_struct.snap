---
source: shapely/tests/derive.rs
expression: "Blah::shape()"
snapshot_kind: text
---
[1;33mBlah[0m (size: [1;34m32[0m, align: [1;35m8[0m)
  [1;32mfoo[0m: 
    [1;33mu32[0m (size: [1;34m4[0m, align: [1;35m4[0m)
    [1;36mScalar:[0m [1;33mU32[0m
  [1;32mbar[0m: 
    [1;33mString[0m (size: [1;34m24[0m, align: [1;35m8[0m)
    [1;36mScalar:[0m [1;33mString[0m
